File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.cc
#include <iostream>
#include <queue>
#include <algorithm>
#include <vector>
using namespace std;

typedef struct node{
    int node;
    int cost;
}Node;

struct compare{
    bool operator()(Node a, Node b){
        return a.cost > b.cost;
    }
};

int solution(vector<vector<int>>& map, vector<int>& dp, int start, int end){


    priority_queue<Node, vector<Node>, compare> nextNode;

    dp[start] = 0;
    nextNode.push({start,0});

    while(nextNode.empty() != true){

        Node curNode = nextNode.top(); nextNode.pop();

        int node = curNode.node;
        int cost = curNode.cost;

        for(int i = 0; i < dp.size(); i++){

            if(map[node][i] == -1)
                continue;

            if(dp[i] > map[node][i] + cost){
                dp[i] = map[node][i] + cost;
                nextNode.push({i,dp[i]});
            }
        }

    }

    return dp[end];
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int n,m;
    cin >> n;
    cin >> m;

    vector<int> dp(n,2100000000);
    vector<vector<int>> map(n,vector<int>(n,-1));

    for(int i = 0; i < m ; i++){
        int from, to, weight;
        cin >> from >> to >> weight;
        if(map[from-1][to-1] == -1)
            map[from-1][to-1] = weight;
        else
            map[from-1][to-1] = min(map[from-1][to-1],weight);
    }

    int start, end;
    cin >> start >> end;

    cout << solution(map,dp,start-1,end-1);
}

File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {

	static class Node{
	    int node;
	    int cost;
	    
	    public Node(int node, int cost) {
	    	this.node = node;
	    	this.cost = cost;
	    			
	    }
	}

	static int[][] map;
	static int[] dp;
	
	static int N,M;

	public static int solution(int start, int end){


	    PriorityQueue<Node> nextNode = new PriorityQueue<>(
	    		(x1,x2) -> {return x1.cost - x2.cost;});
	
	    dp[start] = 0;
	    nextNode.add(new Node(start,0));
	
	    while(nextNode.isEmpty() != true){
	
	        Node curNode = nextNode.poll();
	
	        int cur = curNode.node;
	        int cost = curNode.cost;
	        
	        for(int next = 0; next < N; next++) {
	        	
	        	if(map[cur][next] > 100000)
	        		continue;
	        	
	            if(dp[next] > map[cur][next] + cost){
	                dp[next] = map[cur][next] + cost;
	                nextNode.add(new Node(next, dp[next]));
	            }
	        }
	
	    }
	
	    return dp[end];
	}

	public static void main(String args[]) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in)); 
		StringTokenizer st;
		
		N = Integer.parseInt(br.readLine().trim());
		M = Integer.parseInt(br.readLine().trim());
		 
		dp = new int[N];
		map = new int[N][N];
	
		Arrays.fill(dp, 100000001);
		for(int i = 0; i < N; i++) 
			Arrays.fill(map[i], 100001);
		
	    for(int i = 0; i < M ; i++){
	    	st = new StringTokenizer(br.readLine());
	        int from = Integer.parseInt(st.nextToken()) - 1; 
	        int to = Integer.parseInt(st.nextToken()) - 1;
	        int cost = Integer.parseInt(st.nextToken());	  
	        map[from][to] = Math.min(map[from][to],cost);
	    }

	    st = new StringTokenizer(br.readLine());
	    int start = Integer.parseInt(st.nextToken()) - 1;
	    int end = Integer.parseInt(st.nextToken()) - 1;
	    System.out.println(solution(start, end));
	}
}